enhancements/olm/safe-cluster-upgrades.md

```diff
@@ -0,0 +1,270 @@
+---
+title: Safe OpenShift Upgrades with Operators 
+authors:
+  - "@awgreene"
+reviewers:
+  - TBD
+  - "@kevinrizza"
+approvers:
+  - "@ecordell"
+  - "@spadgett"
+creation-date: 2021-01-11
+last-updated: 2021-01-19
+status: provisional
+see-also:
+  -  N/A
+replaces:
+  - N/A
+superseded-by:
+  - N/A
+---
+
+# Safe OpenShift Upgrades with Operators
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Definitions
+
+**OLM Managed Operators**: Operators that are installed, managed, and upgraded by the [Operator Lifecycle Manager (OLM)](https://github.com/operator-framework/operator-lifecycle-manager) project.
+
+**Minor OpenShift Upgrade**: When the OpenShift Cluster is upgraded to the next minor version, for example the upgrade from version 4.6 to 4.7. See [Semantic Versioning](https://semver.org/) for more details.
+
+## Summary
+
+OpenShift Cluster Admins introduce new services to their clusters in the form of OLM Managed Operators. When performing a cluster upgrade, there is no guarantee that these operators will continue to run on the upgraded cluster.
+
+OLM must make an effort to prevent cluster upgrades when it is possible to determine that one or more installed operators will not run on the next version of OpenShift.
+
+## Motivation
+
+The primary purpose of this enhancement is to define how OpenShift can protect existing services introduced by OLM Managed Operators when performing a Minor OpenShift upgrade.
+
+### Goals
+
+- Prevent Minor OpenShift Upgrades  when an installed operator declares that it does not support the next ocp minor version
+- Warn the Cluster Admin when an installed operator does not explicitly declare support for the next ocp minor version
```

Comment:
it seems like we should also prevent installation of operators that specify a maxKube/OCP version that is lower than the current cluster version.  Can we add that as a goal here? (similar to what minKubeVersion does)

Reply:
Do we want to do that though? I understand the reasoning of minkubeversion because it is easy to be explicit about, but I could imagine a world where an operator is just not maintained anymore and the maxkubeversion isn't accurate but a user really wants to install it, and we do not have a way today of overriding a constraint like that.

Reply:
> we do not have a way today of overriding a constraint like that.

i think that's the actual problem, right?  We need to start introducing ways for admins to override these blocking requirements.

Reply:
This enhancement is focused on preventing unsafe cluster upgrades, not preventing installation of operators onto a OpenShift version that is high than their defined max OpenShift version. I would rather not dilute this enhancement by expanding its scope. If this feature is needed, could we create an enhancement that focuses on expanding the resolver to include it?

Reply:
> This enhancement is focused on preventing unsafe cluster upgrades, not preventing installation of operators onto a OpenShift version that is high than their defined max OpenShift version. 

I have mixed feelings.  I agree that's what this EP is focused on, but i wonder if the EP *should* be focused on "adding a maxocpversion metadata" and all that implies (blocking upgrades, blocking operator installs).

I'm generally happy with iterative development and i don't want to make it harder to get this landed than it needs to be, but it feels weird/confusing to introduce a "maxocp/kubeversion" field for operator authors to specify that doesn't actually enforce that restriction on install, especially when that is exactly what the existing minKubeVersion api field does.  It's also slightly problematic in my mind to add future restrictions based on that field after operator teams start adopting it (we generally try to avoid changing the behavior of apis).

Net being:
1) it can certainly be a separate EP
2) but we need to think long and hard about what it means if we only land the implementation of one of those EPs and not the other, when we introduce the api.

From an implementation perspective, am I naive to think it's pretty trivial to add this constraint to the resolver?  Adding the ability for admins to override the constraint i'm sure is harder, but i don't think we need that on day 1.

Reply:
You are right to insinuate that implementing the constraint is trivial, but the override is a much more hotly debated and nebulous thing that will require significantly more work. I think you do have a good point that introducing this without actually blocking is going to be an unexpected experience for operator devs though.

IMO I think it's reasonable to include that constraint in this enhancement proposal and deal with how to override it in a future enhancement. We aren't actually going to see too many folks with an explicit need to override this until at least one or two releases after we even ship a feature like this.

Reply:
>  We aren't actually going to see too many folks with an explicit need to override this until at least one or two releases after we even ship a feature like this.

yeah, that was my thinking in saying it'd be ok to have the constraint w/o a way to override, initially.